File: Java/src/Question_1_1.java
import static org.junit.Assert.*;
import org.junit.Test;

// Question 1.1:
// Implement an algorithm to determine if a string has all unique characters.
// What if you cannot use additional data structures?

// Assuming that ASCII characters are going to be received in the input string.
public class Question_1_1 {
	static boolean hasUniqueChars(String text) {
		// If more than 256 chars in the input string one of them is repeated.
		if (text.length() > 256)
			return false;

		// Array of booleans to detect duplicates in the text.
		boolean[] charArray = new boolean[256];

		// Get each one of the chars in the text, in order to do this the string
		// has to be converted into a char array to get individual characters.
		for (char c : text.toCharArray()) {
			// If the char was already found, return false.
			if (charArray[c])
				return false;
			// Mark the char as found in the array.
			charArray[c] = true;
		}

		// If no repetitions found, return true.
		return true;
	}

	// Test cases to check the solution of the question, these will use JUnit
	
	// Non repeated characters in the text.
	@Test
	public void NonRepeatedChars() {
		String text = "abcdefg";
		assertTrue(hasUniqueChars(text));
	}

	// Repeated characters in the text.
	@Test
	public void RepeatedChars() {
		String text = "abcdefgabcdefg";
		assertFalse(hasUniqueChars(text));
	}

	// Empty text.
	@Test
	public void EmptyInputString() {
		String text = "";
		assertEquals(true, hasUniqueChars(text));
	}
}

File: JavaScript/Question_1_1.js
// Question 1.1:
// Implement an algorithm to determine if a string has all unique characters.
// What if you cannot use additional data structures?

// Method to check is a string has repeated characters, assuming that the input text is ASCII.
function hasUniqueChars(text)
{
    // Check if the text has more than 256 characters, meaning that at least one is repeated.
    if (text.length > 256) {
        return false;
    }
    
    // Create an array of boolean values of 256 elements and put initialize them with zeros.
    var charArray = Array.apply(null, Array(256)).map(Boolean);
    
    // Iterate in the elements of the text, will get the index of the text instead of the char.
    for (i in text) {
        // If the character was already found return false.
        if(charArray[text[i]]) {
            return false;
        }
        // Mark the character in the array.
        charArray[text[i]] = true;
    }
    
    // If no repeated characters were found, return true to mark that the text has unique chars.
    return true;
}

// Using the QUnit framework to do the testing of the function.

// Create a module to put the tests togheter on the result page.
QUnit.module( "Question_1_1" );

// Test for non repeated characters in the text.
QUnit.test("NonRepeatedChars", function(assert) {
    var text = "abcdefg";
    assert.ok(hasUniqueChars(text));
});

// Test for repeated characters in the text.
QUnit.test("RepeatedChars", function(assert) {
    var text = "abcdefgabcdefg";
    assert.ok(!hasUniqueChars(text), false, "Success");
});

// Empty text test.
QUnit.test("EmptyInputString", function(assert) {
    var text = "";
    assert.ok(hasUniqueChars(text));
});